spec/abstract/InvocationFuns_A.thy

```diff
@@ -0,0 +1,385 @@
+(*
+ * Copyright 2014, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(*
+Monad instantiations, handling of faults, errors, and interrupts.
+*)
+
+chapter "Basic Kernel and Exception Monads"
+
+theory InvocationFuns_A
+imports Ipc_A
+begin
+
+section \<open>Allow preemption at this point.\<close>
+definition
+  preemption_point :: "(unit,'z::state_ext) p_monad" where
+ "preemption_point \<equiv> doE liftE $ do_extended_op update_work_units;
+                         OR_choiceE (work_units_limit_reached)
+                           (doE liftE $ do_extended_op reset_work_units;
+                                irq_opt \<leftarrow> liftE $ do_machine_op (getActiveIRQ True);
+                                case_option (do
```

Comment:
Nitpick: were you intending to put the first line of the `do` block on a new line? I'd prefer it on the same line with the `do` and `od` indented the same number of spaces, but I know there are other thoughts on this

Reply:
Hmm. I think I do like what you are saying, too, but when the block surrounded by `do` and `od` is small, it feels to spread a bit too thinly sometimes, and also when the block is in the branches of `if-then-else`, I would prefer to see the `if-then-else` structure clearly than the `do`-block (also because the `do`-`od` demarcation is redundant in this case here). Something like that is what I was thinking in the background here. But not with any strong opinion, really.

Reply:
So, this is now ready for you to start the proof :)